File: packages/contracts/src/artifacts.ts
import { z } from "zod"

import {
  NonEmptyTextSchema,
  RevisionSchema,
  Sha256Schema,
  UtcTimestampSchema,
  Uuid7Schema,
} from "./common"

export const ArtifactCreateSchema = z
  .strictObject({ project_id: Uuid7Schema, name: NonEmptyTextSchema })
  .readonly()
export const ArtifactSchema = z
  .strictObject({
    id: Uuid7Schema,
    org_id: Uuid7Schema,
    project_id: Uuid7Schema,
    name: NonEmptyTextSchema,
    created_at: UtcTimestampSchema,
  })
  .readonly()
export const ArtifactVersionCreateSchema = z
  .strictObject({
    base_version: RevisionSchema,
    checksum_sha256: Sha256Schema,
    size_bytes: z.int().min(0),
    media_type: NonEmptyTextSchema,
    producing_run_id: Uuid7Schema,
  })
  .readonly()
export const ArtifactVersionSchema = z
  .strictObject({
    id: Uuid7Schema,
    org_id: Uuid7Schema,
    artifact_id: Uuid7Schema,
    version: RevisionSchema,
    checksum_sha256: Sha256Schema,
    size_bytes: z.int().min(0),
    media_type: NonEmptyTextSchema,
    producing_run_id: Uuid7Schema,
    created_at: UtcTimestampSchema,
  })
  .readonly()
export const ReviewCreateSchema = z
  .strictObject({
    source_run_id: Uuid7Schema,
    artifact_version_ids: z.array(Uuid7Schema).default([]).readonly(),
    execution_ids: z.array(Uuid7Schema).default([]).readonly(),
  })
  .refine(
    (review) => review.artifact_version_ids.length > 0 || review.execution_ids.length > 0,
    "Review requires Artifact Version or Execution pins",
  )
  .readonly()
export const ReviewSchema = z
  .strictObject({
    id: Uuid7Schema,
    org_id: Uuid7Schema,
    source_run_id: Uuid7Schema,
    run_id: Uuid7Schema,
    status: z.enum(["queued", "running", "completed", "failed"]),
    created_at: UtcTimestampSchema,
  })
  .readonly()
export const ExportCreateSchema = z
  .strictObject({ artifact_version_ids: z.array(Uuid7Schema).min(1) })
  .readonly()
export const ExportSchema = z
  .strictObject({
    id: Uuid7Schema,
    org_id: Uuid7Schema,
    status: z.enum(["queued", "running", "completed", "failed"]),
    artifact_version_ids: z.array(Uuid7Schema).readonly(),
    created_at: UtcTimestampSchema,
  })
  .readonly()

export const ProviderConnectionCreateSchema = z
  .strictObject({
    adapter_id: z.enum([
      "openai_codex",
      "anthropic_claude_code",
      "xai_grok_build",
      "moonshot_kimi_code",
    ]),
  })
  .readonly()
export const ProviderConnectionSchema = z
  .strictObject({
    id: Uuid7Schema,
    org_id: Uuid7Schema,
    requester_user_id: Uuid7Schema,
    adapter_id: z.enum([
      "openai_codex",
      "anthropic_claude_code",
      "xai_grok_build",
      "moonshot_kimi_code",
      "zai_glm",
    ]),
    status: z.enum(["pending", "healthy", "reauth_required", "revoked", "unsupported_auth"]),
    created_at: UtcTimestampSchema,
  })
  .readonly()
export const DeletionCreateSchema = z.strictObject({ project_id: Uuid7Schema }).readonly()
export const DeletionRequestSchema = z
  .strictObject({
    id: Uuid7Schema,
    org_id: Uuid7Schema,
    status: z.enum(["queued", "running", "completed", "held", "failed"]),
    created_at: UtcTimestampSchema,
  })
  .readonly()
export const LegalHoldStatusSchema = z
  .strictObject({ org_id: Uuid7Schema, active: z.boolean(), updated_at: UtcTimestampSchema })
  .readonly()

export type ArtifactVersion = z.infer<typeof ArtifactVersionSchema>
export type Review = z.infer<typeof ReviewSchema>

File: packages/contracts/python/science_workbench_contracts/artifacts.py
from __future__ import annotations

from typing import Annotated, Literal, Self

from pydantic import Field, model_validator
from pydantic_core import PydanticCustomError

from .common import (
    ContractModel,
    NonEmptyText,
    Revision,
    Sha256,
    UtcTimestamp,
    Uuid7,
)


class ArtifactCreate(ContractModel):
    project_id: Uuid7
    name: NonEmptyText


class Artifact(ContractModel):
    id: Uuid7
    org_id: Uuid7
    project_id: Uuid7
    name: NonEmptyText
    created_at: UtcTimestamp


class ArtifactVersionCreate(ContractModel):
    base_version: Revision
    checksum_sha256: Sha256
    size_bytes: Annotated[int, Field(ge=0)]
    media_type: NonEmptyText
    producing_run_id: Uuid7


class ArtifactVersion(ContractModel):
    id: Uuid7
    org_id: Uuid7
    artifact_id: Uuid7
    version: Revision
    checksum_sha256: Sha256
    size_bytes: Annotated[int, Field(ge=0)]
    media_type: NonEmptyText
    producing_run_id: Uuid7
    created_at: UtcTimestamp


class ReviewCreate(ContractModel):
    source_run_id: Uuid7
    artifact_version_ids: tuple[Uuid7, ...] = ()
    execution_ids: tuple[Uuid7, ...] = ()

    @model_validator(mode="after")
    def validate_pins(self) -> Self:
        if not self.artifact_version_ids and not self.execution_ids:
            error = PydanticCustomError(
                "missing_review_pins",
                "Review requires Artifact Version or Execution pins",
            )
            raise error
        return self


class Review(ContractModel):
    id: Uuid7
    org_id: Uuid7
    source_run_id: Uuid7
    run_id: Uuid7
    status: Literal["queued", "running", "completed", "failed"]
    created_at: UtcTimestamp


class ExportCreate(ContractModel):
    artifact_version_ids: Annotated[tuple[Uuid7, ...], Field(min_length=1)]


class Export(ContractModel):
    id: Uuid7
    org_id: Uuid7
    status: Literal["queued", "running", "completed", "failed"]
    artifact_version_ids: tuple[Uuid7, ...]
    created_at: UtcTimestamp


class ProviderConnectionCreate(ContractModel):
    adapter_id: Literal[
        "openai_codex",
        "anthropic_claude_code",
        "xai_grok_build",
        "moonshot_kimi_code",
    ]


class ProviderConnection(ContractModel):
    id: Uuid7
    org_id: Uuid7
    requester_user_id: Uuid7
    adapter_id: Literal[
        "openai_codex",
        "anthropic_claude_code",
        "xai_grok_build",
        "moonshot_kimi_code",
        "zai_glm",
    ]
    status: Literal[
        "pending", "healthy", "reauth_required", "revoked", "unsupported_auth"
    ]
    created_at: UtcTimestamp


class DeletionCreate(ContractModel):
    project_id: Uuid7


class DeletionRequest(ContractModel):
    id: Uuid7
    org_id: Uuid7
    status: Literal["queued", "running", "completed", "held", "failed"]
    created_at: UtcTimestamp


class LegalHoldStatus(ContractModel):
    org_id: Uuid7
    active: bool
    updated_at: UtcTimestamp
